### PR TITLE
Add Dependency Injection support

### LINKED ISSUE #148: Error when migrating to 4.0
Hi, Exiting news about the 4.0 version! 

I have got this code:
https://dotnetfiddle.net/5iSl8A
That is currently running on my production with Ncalc 3.11 and working beautifully. 

I want to migrate to 4.0 to enjoy your hard work for efficiency and  the new getParameters function. 

But, when doing so, I get few errors and warnings in my code: 

but unfortunately when trying to migrate to 4.0 I get few errors: 
Compilation error (line 91, col 27): The type or namespace name 'EvaluationVisitor' could not be found (are you missing a using directive or an assembly reference?)

Compilation error (line 93, col 20): The type or namespace name 'EvaluateOptions' could not be found (are you missing a using directive or an assembly reference?)

Compilation error (line 93, col 46): The name 'EvaluateOptions' does not exist in the current context

Can you help me figure it out?

### LINKED ISSUE #154: Dependency Injection
**Is your feature request related to a problem? Please describe.**
I want use `IMemoryCache` at the cache, but don't want to create a dependency in NCalc with this.

**Describe the solution you'd like**
Create a `NCalc.DependencyInjection` package with a dependency on `Microsoft.Extensions.DependencyInjection`

At `Program.cs`:
```
builder.Services.AddNCalc()
.WithCache<CustomCacheImplementation>()
.WithParser<CustomParser>
.WithEvaluationVisitor<CustomEvaluationVisitor>();

builder.Services.AddTransient<MyService>():
```

Usage:
```
using NCalc;

namespace JJMasterData.NCalc;

public class MyService(IExpressionFactory expressionFactory)
{
  
    public object? Evaluate(string expression)
    {
        var expression = expressionFactory.Create(replacedExpression);
 
        return expression.Evaluate();
    }
 
}
```

**Describe alternatives you've considered**
Inheriting `Expression`, but feels over-engineering, I don't like currently on how we create custom `EvaluationVisitor`s.

**Additional context**
My team on JJMasterData also want to use `{}` as identifier, we currently pre-parse the expression to find these identifiers, a `ILogicalExpressionParser` would solve the problem.

### LINKED ISSUE #163: Docs: Document the `ExpressionOptions` enum values
image

### PR DESCRIPTION
**Pull request by gumbarros**
This closes #154 
This closes #163 
This closes #148

### CODE REVIEW

**Comment by gumbarros** [comment]
@Bykiev ready for review. Still needs docs, it's 23pm here, time for some sleep :)

File: src/NCalc/ExpressionContext.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    }
+    
+    public static implicit operator ExpressionContext(ExpressionOptions options)
+    {
+        return new ExpressionContext { Options = options };
+    }
+    
+    public static implicit operator ExpressionContext(CultureInfo cultureInfo)
+    {
+        return new ExpressionContext { CultureInfo = cultureInfo };
```
Comment (gumbarros) [inline src/NCalc/ExpressionContext.cs:26]: Just context for reviewers, this prevent a breaking change in the `Expression` constructors (the proof is the unit tests are untouched)

**Comment by Bykiev** [comment]
Great job, I'll review it carefully at evening. These changes seems too complex and breaking, I'd say it's for NCalc v5. What do you think?

**Comment by gumbarros** [comment]
> Great job, I'll review it carefully at evening. These changes seems too complex and breaking, I'd say it's for NCalc v5. What do you think?

I'm in doubt, if we were to follow semver, it would really be v5, but I consider the breaking change nearly irrelevant, I think is very hard someone using `LogicalExpressionVisitor`, and in the scenario of using, just a I prefix solve the breaking change.

Ohhh, wait, typing this I just discovered how we workaround and not have breaking changes, maybe I can just add `LogicalExpressionVisitor` again and implement the interface.

So, without breaking changes, I vote in 4.1.

**Comment by gumbarros** [comment]
Now No Breaking Changes, in a v5 we can remove the class (also added `ObsoleteAttribute`) to help in a clean migration any consumer of this class. (very hard to use, but breaking change prevented)

**Review by Bykiev — COMMENTED** [review]
Looking good, can you please add some plugin tests? Also we need to update the docs. When you'll merge, can you please create a separate branch, not in the master. I think this feature will be ready for production in NCalc v4.2 and v 4,1 will include bug fixes and a new features - semicolon as argument delimeter an compare with null parameters

**Comment by gumbarros** [comment]
@Bykiev I made unit tests for the MemoryCache, I'm going to [sextou](https://www.google.com/search?q=sextou+meaning&oq=sextou+meaning&gs_lcrp=EgZjaHJvbWUyCQgAEEUYORiABDIHCAEQABiABDIICAIQABgWGB4yCAgDEAAYFhgeMg0IBBAAGIYDGIAEGIoFMg0IBRAAGIYDGIAEGIoFMg0IBhAAGIYDGIAEGIoFMgoIBxAAGIAEGKIE0gEIMTc3MmowajGoAgCwAgA&sourceid=chrome&ie=UTF-8) here, tomorrow I will finish the docs

**Comment by gumbarros** [comment]
@Bykiev ready to merge. But I vote in a single release of 4.1, it will be maintenance hell of both 4.1 an 4.2 with different feats I think. This PR does not introduce any breaking changes, and I already wants to use this at my project.

**Comment by gumbarros** [comment]
Finished my self-review, LGTM now.

**Comment by Bykiev** [comment]
Agreed, as there is no any breaking changes, let's include it in v4.1. Don't see any issues, can be merged

**Pull request merged**